core: reset controller driver interface

etienne-lms opened the pull request:
This P-R proposes a reset controller driver interface and means to register in the DT driver provider framework ~~and some embedded tests~~:
* commit "_drivers: add reset controller framework_"
* ~~commit "_core: dt_driver: test reset-controller probing_"~~

~~This P-R reuses other changes alrerady under in others P-R:~~
~~* commit "_core: dt_driver: pass driver type to clk_get_provider_by_*()_" (from [P-R 4988](https://github.com/OP-TEE/optee_os/pull/4988)) is needed for provider to possibly differentiate providers relate to a same device node in the embedded DT.~~
~~* commits "core: dt_driver: return TEE_ERROR_DEFER_DRIVER_INIT if no provider",
"drivers: clk: add CFG_DRIVERS_CLK_EARLY_PROBE" and "core: dt_driver: drivers to test probe deferral"
(from [P-R 4995](https://github.com/OP-TEE/optee_os/pull/4995)) are foundation of DT driver probing tests, needed to add test on reset controller drivers probing.~~

etienne-lms commented:
I updated the P-R, removing the embedded test part that will come later.
Remains a single commit that propose a generic API for supporting reset controller in OP-TEE.

etienne-lms commented:
Rebased on master.

Any comment on that. Please tell me is this seems ok as a first step for reset controller support, or if it rather looks like a bad idea or API design.

Checkpatch `CHECK` report is not blocking IMO:
```
CHECK: Macro argument reuse '_ops' - possible side-effects?
#202: FILE: core/include/drivers/rstctrl.h:72:
+#define RSTCTRL_OPS_IS_VALID(_ops) \
+	((_ops)->assert_level && (_ops)->deassert_level)
```

jenswi-linaro commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,217 @@
+		.probe = __probe, \
+	}
+
+#define RSTCTRL_OPS_IS_VALID(_ops) \
```
Any reason why this is a macro instead of an inline function? The latter should take care of the checkpatch issue.

etienne-lms replied:
Right, I'll move to an inline function. (I should have done it seen checkpatch report)

jenswi-linaro commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,217 @@
+ */
+struct rstctrl {
+	const struct rstctrl_ops *ops;
+	uintptr_t priv;
```
Do we need this field? I guess it wouldn't if we intended to include this struct in a larger device specific struct.

etienne-lms replied:
Indeed, could be removed.

etienne-lms replied:
do we restore this generic private ref for the driver?

In current series state, it is up to the platform reset controller driver to bind the struct address to its local reset line reference. 

Example: planned for [_stm32_reset.c_ (ephemeral link)](https://github.com/etienne-lms/optee_os/commit/2e12ef74f202562574fff9dc42009d3145b73d67?diff=unified#diff-1ceec50dd256ff8eba2db4dd1c4f52dd1ba4214b40b3e5af0ea89966c4748d1bR22)
```c
/* Exposed rstctrl instance */
struct stm32_rstline {
	unsigned int id;
	struct rstctrl rstctrl;
	SLIST_ENTRY(stm32_rstline) link;
};

static struct stm32_rstline *to_rstline(struct rstctrl *rstctrl)
{
	assert(rstctrl);

	return container_of(rstctrl, struct stm32_rstline, rstctrl);
}
```

Adding `priv` in `struct rstctrl` increases the struct size, may be be easier for platforms willing to save a reference (local pointer or numeric identifer.
Note the numeric identifer must be set by the provider driver: it may be the DT binding Id for 1 cell `#reset-cells` providers, it may need a conversion when DT binding rely or more IDs (`#reset-cells = 2;`).

jenswi-linaro commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,217 @@
+struct rstctrl {
+	const struct rstctrl_ops *ops;
+	uintptr_t priv;
+	bool exclusive;
```
How will this field be used? Is it needed for all instances?

etienne-lms replied:
In case a reset controller is shared by several devices, one may request exclusive access to the controller when using the related resource and release exclusive access once resource is no more used. Likely needed only for specific shared reset controllers. I've added here to have a generic structure.

etienne-lms replied:
> How will this field be used? 

Sorry, I didn't answer that question. It is used through `rstctrl_{get|put}_exclusive()`.

jenswi-linaro replied:
Thanks

jenswi-linaro commented on core/drivers/rstctrl/rstctrl.c:
```diff
@@ -0,0 +1,63 @@
+{
+	assert(rstctrl->exclusive);
+
+	rstctrl->exclusive = false;
```
I would prefer if this was assigned using a `WRITE_ONCE()` macro. We don't have that macro yet, but it should be easy to implement using `__compiler_atomic_store()`

jenswi-linaro commented on core/drivers/rstctrl/rstctrl.c:
```diff
@@ -0,0 +1,63 @@
+							   nodeoffset, idx,
+							   res);
+
+	return (struct rstctrl *)device;
```
Cast is not needed.

etienne-lms commented:
comments addressed

jenswi-linaro commented:
For "io.h: add WRITE_ONCE macro" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

For the rest:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
Many thanks for the review and feedback.
Review tags applied.

etienne-lms commented:
I've added a fixup commit.

jenswi-linaro commented:
The fixup looks good:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
thanks @jenswi-linaro, i'll squash it.

jforissier commented:
The state of this PR looks good enough to be merged but since it is a RFC perhaps more feedback is needed. @clementleger since you submitted #5028, what do you think? Does this look useful to you? Thanks.

clementleger commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,228 @@
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the subnode containing a 'resets' property
+ * @index: Reset controller index in 'resets' property
```
should be `idx`

etienne-lms replied:
right, thanks.

clementleger commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,228 @@
+ * rstctrl_name - Get a name for the reset level control or NULL
+ *
+ * @rstctrl: Reset controller
+ * Return a pointer to  anem or NULL
```
typo: `anem`

etienne-lms replied:
thanks

clementleger commented on core/drivers/sub.mk:
```diff
@@ -47,5 +47,6 @@ srcs-$(CFG_ZYNQMP_HUK) += zynqmp_huk.c
 subdirs-y += crypto
 subdirs-$(CFG_BNXT_FW) += bnxt
 subdirs-$(CFG_DRIVERS_CLK) += clk
+subdirs-$(CFG_DRIVERS_RSTCTRL) += rstctrl
```
Shouldn't `CFG_DRIVERS_RSTCTRL` be documented in `config.mk` ?

etienne-lms replied:
yes it should. i'll add. thanks.

clementleger reviewed (commented):
This support seems a good starting point but the lack of multiple reset line support probably lacks.

etienne-lms commented:
> Driver I submitted is a global reset for the SoC. This RFC targets reset controller which can reset individual device AFAIU.

I fully agree.

> EDIT: I would say that it could be used to reset the chip but then, maybe some additional property should be added (...)
> Or maybe something on top of that could make use of it to reset the chip and this generic code could be used by PSCI for reset for instance.

I would not go that way. Reset controllers are designed to be use by consumers: drivers (using a dt "resets = " property) over a reset controller provider driver.  PSCI does not get system or CPU reset controllers from DT, not for the CPU of the executing REE/TEE.

> This support seems a good starting point but lacks support of multiple reset line per controller.

Either consumer driver store both reset controller instance reference (pointer) and a controller ID.
=> `rstctrl_[de]assert()` gets an extra argument.
Or the reset controller is 1 instance per reset line: this proposal.
=> current way, diver use private means to bind the struct rstctrl * reference to a target line.
Alternate to current state: restore `priv` field in `struct rstctrl` to hold the reset line ref in the reset controller driver.

etienne-lms commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,228 @@
+	 *
+	 * @rstctrl: Reset control reference
+	 * @id: Identifier for the reset level control in the reset controller
+	 * @to_ms: Timeout in microsencond or RSTCTRL_NO_TIMEOUT, may be ignored
```
> id and to_ms are not parameters

Right, I described them here since common arguments to the operators.
You would prefer another way to describe them?

> BTW, I do not see where you can specify the reset line to use. 

Consumer calls `rstctrl_dt_get_by_index()` (or `rstctrl_dt_get_by_name()`, as clocks) and gets a reset line, I.e. a `struct rstctrl` pointer.
It then calls `rstctrl_assert()` to request request.
It may (often) need to call `rstctrl_deassert()` when the line shall be manually released.

Should rstctrl_dt_get_by_index() return a rstctrl reference + a line ID reference, to be used in rstctrl_*() API functions?

etienne-lms commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,228 @@
+
+/**
+ * RSTCTRL_DECLARE - Declare a reset controller driver with a single
+ * DTT compatible string.
```
will fix. thanks.

etienne-lms commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,228 @@
+}
+
+/**
+ * rstctrl_dt_get_by_idx - Get a reset controller at a specific index in
```
I want to rename this to: `dt_get_rstctrl_by_index()` with get `TEE_result` return value:
```c
TEE_Result dt_get_rstctrl_by_index(const void *fdt, int nodeoffset,
				   unsigned int index,
				   struct rstctrl **rstctrl);
```
It will look better in caller's instruction block:
```c
	struct rstctrl *rstctrl_ref = NULL;

	res = dt_get_rstctrl_by_index(fdt, node, index, &rstctrl_ref);
	if (res)
		return res;
	res = rstctrl_get_exclisive(rstctrl_ref);
	if (res)
		return res;
	(...)
```

I'll propose the same for clk: 
```c
TEE_Result dt_get_clk_by_index(const void *fdt, int nodeoffset,
			       unsigned int index, struct clk **clk);
TEE_Result dt_get_clk_by_name(const void *fdt, int nodeoffset,
			      chanr *name, struct clk **clk);
```

clementleger replied:
Looks good to me for the clock functions.

etienne-lms reviewed (commented):
Many thanks for the feedback and review.

etienne-lms commented:
@clementleger, regarding `priv` field in the reset controller structure (`struct rstctrl`), would you you prefer the field inside the structure or let platform driver include `struct rstcltr` inside its own instance structure (and use `container_of()` like means, as in current P-R state)?

clementleger commented:
Maybe letting the platform driver decides where to put it will allow more flexibility ? `container_of()` will allow all of them to get it easily and it will also allow all drivers to put this in the memory they want.

etienne-lms commented:
Do I understand the current proposal is fine, without a `priv` field in `struct rstctrl`?
```c
struct rstctrl {
	const struct rstctrl_ops *ops;
	bool exclusive;
};
```

clementleger commented:
> Do I understand the current proposal is fine, without a `priv` field in `struct rstctrl`?
> 
> ```c
> struct rstctrl {
> 	const struct rstctrl_ops *ops;
> 	bool exclusive;
> };
> ```

Yes !

etienne-lms commented:
ok, thanks to confirm. I'll update this P-R and create one for the clock APIs.

etienne-lms commented:
Comments addressed.

clementleger commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,227 @@
+}
+
+/**
+ * rstctrl_dt_get_exclusive - Get exclusive control or target
```
typo: `of` instead of `or`

etienne-lms replied:
yes

clementleger commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2021, Linaro Limited
+ *
+ * Self contained reset controller framework in rstctrl.h
```
Is it still self contained in the `.h` since there is a `rstctrl.c` file ?

clementleger commented on mk/config.mk:
```diff
@@ -727,6 +727,13 @@ CFG_DRIVERS_CLK ?= n
+
+# When enabled, CFG_DRIVERS_RSTCTRL embeds a reset controller framework in
+# OP-TEE core to provide reset controls on subsystems of the devices.
+CFG_DRIVERS_RSTCTRL ?= n
```
Don't you have a dependency on `CFG_DT` ?

etienne-lms replied:
There should not be but for the rstctrl_dt_get_by_`*()` API functions.
I'll check that it build ok with and without `CFG_DT`.

etienne-lms commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,227 @@
+/**
+ * rstctrl_dt_get_exclusive - Get exclusive control or target
+ *
+ * @rstctrl: Reference to reset level control
```
ok

etienne-lms commented on core/include/drivers/rstctrl.h:
```diff
@@ -0,0 +1,227 @@
+/**
+ * rstctrl_put_exclusive - Release exclusive access to target
+ *
+ * @rstctrl: Reference to reset level control
```
ok

etienne-lms commented:
Comments addressed.
I've rebased the series with its fixup commits to latest master to check build when `CFG_DT` is disabled.
I've added some more fixup commits to address latest review comments.

jforissier commented:
For commit "mk/config.mk: fix layout":
s/a assigned/are assigned/, then
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

etienne-lms commented:
Updated the series with reviewed commits + non-reviewed changes (API function change) as a fixup commit.

etienne-lms commented:
CI seems to fails on a [python errors](https://dev.azure.com/OPTEE/optee_os/_build/results?buildId=2068&view=logs&j=705748f3-7146-5e86-79af-1c0266d20a8c&t=d3458a7d-1a81-5639-8e63-8927d807e53a&l=24447) in the build log?
```
ImportError: cannot import name 'SystemDependency' from 'mesonbuild.dependencies.base' (snip)
```

jforissier commented:
> CI seems to fails on a [python errors](https://dev.azure.com/OPTEE/optee_os/_build/results?buildId=2068&view=logs&j=705748f3-7146-5e86-79af-1c0266d20a8c&t=d3458a7d-1a81-5639-8e63-8927d807e53a&l=24447) in the build log?
> 
> ```
> ImportError: cannot import name 'SystemDependency' from 'mesonbuild.dependencies.base' (snip)
> ```

Yes, I think it is just a matter of rebuilding the Docker image (Buildroot changed and this required cleaning or rebuilding from scratch). I'm on it.

etienne-lms commented:
thanks @jforissier.

etienne-lms commented:
@jenswi-linaro, sorry to bother. You've acked the main commit but from discussions, I changed an API function prototype and a fixup commit remains un-reviewed.
Could you ack the fixup if looking good to you, so that i could get the whole merged?

jenswi-linaro commented on core/include/drivers/rstctrl.h:
```diff
@@ -141,30 +141,42 @@ static inline bool rstctrl_ops_is_valid(const struct rstctrl_ops *ops)
-						    int nodeoffset,
-						    unsigned int idx,
-						    TEE_Result *res)
+static inline TEE_Result rstctrl_dt_get_by_index(const void *fdt,
```
Mindor: How about keeping this name as `rstctrl_dt_get_by_idx()` instead? Same for the idx parameter.

etienne-lms replied:
It is rather for consistency with the change made in clk_dt_get_by*() that I change here `idx` to plain `index`.

jenswi-linaro reviewed (commented):
Only a minor comment on the fixup. With that fixed or not:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
Thanks for the review.
Tags applied.

jforissier commented:
@clementleger any more feedback?

clementleger commented:
Nope, that is ok for me.

etienne-lms commented:
I need to rebase since introduction of macro `DEFINE_DT_DRIVER()`
